src/AppScopeTrait.php

```diff
@@ -59,4 +66,43 @@ trait AppScopeTrait
      * @var array
      */
     public $unique_hashes = [];
+
+    /**
+     * To be removed in Jan 2021.
+     */
+    private function assertNoDirectAssignment(): void {
+        if ($this->app !== null) {
+            throw new Exception('App can not be assigned directly');
+        }
+    }
+
+    public function issetApp(): bool
+    {
+        $this->assertNoDirectAssignment();
+
+        return $this->_app !== null;
+    }
+
+    public function getApp(): App
```

Comment:
impossible to test if there is no App class - do we want this trait really in core? If yes, we can enforce object type and phpdoc App.